File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int foo(vector<int>& A,vector<int>& dp, int i)
    {
        if(i>=size(A)) return 0; 
        if(dp[i] != -1) return dp[i];
        return dp[i]=max(A[i]+foo(A,dp,i+2),foo(A,dp,i+1));
    }
    
    int rob(vector<int>& A) {
        vector<int> dp(size(A),-1);
        return foo(A,dp,0);
    }
        
};

File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        if (n == 1) return nums[0];
        int [] dp =new int[n];
        dp[0]=nums[0];
        dp[1]=Math.max(nums[1],nums[0]);
        for(int i=2;i<nums.length;i++)
        {
            dp[i]=Math.max(dp[i-2]+nums[i],dp[i-1]);
        }
        return dp[n-1];
    }
}